### PR TITLE
Elastic to docs

### PR DESCRIPTION
**Pull request by StarfallProjects**

### CODE REVIEW

File: app/views/layouts/_sidebar.html.erb [hunk 1]
```diff
@@ -49,6 +49,7 @@
           <%= sidebar_link_to "Linux", 'agent/v3/linux' %>
           <%= sidebar_link_to "Docker", 'agent/v3/docker' %>
           <%= sidebar_link_to "AWS", 'agent/v3/aws' %>
+          <%= sidebar_link_to "Elastic CI Stack", 'agent/v3/elastic_ci' %>
```
Comment (yob) [inline app/views/layouts/_sidebar.html.erb:52]: maybe `Elastic CI Stack for AWS`?

In the future we might have CI stacks for alternative cloud providers, and is so they'll probably be different enough to require their own docs page.

Comment (StarfallProjects) [reply app/views/layouts/_sidebar.html.erb]: Done

File: pages/agent/v3/elastic_ci.md.erb [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+By following these conventions you get a scalable, repeatable, and source-controlled CI environment that any team within your organization can use.
+
+## Multiple instances of the stack
+
+If you need different instances sizes and scaling characteristics for different pipelines, you can create multiple stacks. Each can run on a different [Agent queue](/docs/agent/queues), with its own configuration, or even in a different AWS account.
+
+Examples:
+
+* A `docker-builders` stack that provides always-on workers with hot Docker caches (see [Optimizing for slow Docker builds](#optimizing-for-slow-docker-builds))
+* A `pipeline-uploaders` stack with tiny, always-on instances for lightning fast `buildkite-agent pipeline upload` jobs.
```
Comment (plaindocs) [inline pages/agent/v3/elastic_ci.md.erb:111]: I wonder if @yob could find us links to examples for these other two?

Comment (plaindocs) [reply pages/agent/v3/elastic_ci.md.erb:111]: Or @jayco ?

File: pages/agent/v3/elastic_ci.md.erb [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+
+This means you can scale down to zero when idle, which means you can use larger instances for the same cost.
+
+Metrics are collected with a Lambda function, polling every minute based on the queue the stack is configured with. The autoscaler monitors only one queue.
+
+## Terminating the instance after the job is complete
+
+You can set `BuildkiteTerminateInstanceAfterJob` to `true` to force the instance to terminate after it completes a job. Setting this value to `true` tells the stack to enable `disconnect-after-job` in the `buildkite-agent.cfg` file.
+
+We strongly encourage you to find an alternative to this setting if at all possible. The turn around time for replacing these instances is currently slow (5-10 minutes depending on other stack configuration settings). If you need single use jobs, we suggest looking at our container plugins like `docker`, `docker-compose`, and `ecs`, all which can be found [here](https://buildkite.com/plugins).
```
Comment (plaindocs) [inline pages/agent/v3/elastic_ci.md.erb:126]: Another "we" we should rewrite.

**Review by plaindocs — CHANGES_REQUESTED** [review]
This is awesome! So excited to get this into the main docs site.
I've left a few suggestions inline

File: pages/agent/v3/elastic_ci.md.erb [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+
+If you use [Amazon ECR](https://aws.amazon.com/ecr/) you can set the `ECRAccessPolicy` parameter for the stack to either `readonly`, `poweruser`, or `full` depending on the [access level](http://docs.aws.amazon.com/AmazonECR/latest/userguide/ecr_managed_policies.html) you want your builds to have.
+
+You can disable this in individual pipelines by setting `AWS_ECR_LOGIN=false`.
+
+If you want to log in to an ECR server on another AWS account, you can set `AWS_ECR_LOGIN_REGISTRY_IDS="id1,id2,id3"`.
+
+The AWS ECR options are powered by an embedded version of the [ECR plugin](https://github.com/buildkite-plugins/ecr-buildkite-plugin), so if you require options that aren't listed here, you can disable the embedded version as above and call the plugin directly. See [its README](https://github.com/buildkite-plugins/ecr-buildkite-plugin) for more examples (requires Agent v3.x).
+
+## Versions
```
Comment (plaindocs) [inline pages/agent/v3/elastic_ci.md.erb:146]: Can we call this something better than Versions?  Elastic CI Stack versions? Elastic CI stack Releases?

File: app/views/layouts/_sidebar.html.erb (hunk 1 above)
Comment (plaindocs) [reply app/views/layouts/_sidebar.html.erb]: Shall we rename the file also? Probably best

File: pages/agent/v3/elastic_ci.md.erb [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+
+## Security
+
+This repository hasn't been reviewed by security researchers so exercise caution and careful thought with what credentials you make available to your builds.
+
+Anyone with commit access to your codebase (including third-party pull-requests if you've enabled them in Buildkite) has access to your secrets bucket files.
+
+Also keep in mind the EC2 HTTP metadata server is available from within builds, which means builds act with the same IAM permissions as the instance.
+
+## Development
```
[the remark this answers is not in the document]
Comment (StarfallProjects) [reply pages/agent/v3/elastic_ci.md.erb:246]: As I understood it, they would need this info if tweaking the thing themselves, not just if contributing?

[the remark this answers is not in the document]
Comment (plaindocs) [reply pages/agent/v3/elastic_ci.md.erb]: Probs. @yob is this common? Do you have thoughts on leaving it in the README vs in the docs?

File: pages/agent/v3/elastic_ci.md.erb [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+    agents:
+      queue: elastic
+    parallelism: 75
+```
+
+See [Issue 81](https://github.com/buildkite/elastic-ci-stack-for-aws/issues/81) for ideas on other solutions (contributions welcome!).
+
+## Security
+
+This repository hasn't been reviewed by security researchers so exercise caution and careful thought with what credentials you make available to your builds.
```
[the remark this answers is not in the document]
Comment (yob) [reply pages/agent/v3/elastic_ci.md.erb]: yer, that feels like a phrase that might be OK in the README but not ideal in the official docs.

File: pages/agent/v3/elastic_ci.md.erb (hunk 5 above)
[the remark this answers is not in the document]
Comment (yob) [reply pages/agent/v3/elastic_ci.md.erb]: I reckon the development section could stay in the README and be left out of the official docs.

Regular users of the stack won't need these instructions, and anyone who needs something customised enough to need these instructions will be forking the repo and making changes there.

File: pages/agent/v3/elastic_ci.md.erb (hunk 2 above)
Comment (yob) [reply pages/agent/v3/elastic_ci.md.erb]: I'm not sure if we have any public documentation on these kinds of queues and agent patterns. I suspect we may not.

@jayco is definitely the one to ask! @toolmantim might also know.

Comment (jayco) [reply pages/agent/v3/elastic_ci.md.erb]: Hmm, I don't think we actually have any examples of this. 

I normally point to optimizing-for-slow-docker-builds that you have already linked to.

**Comment by plaindocs** [comment]
@StarfallProjects try switching to this image syntax `<%= image "ec2-asg.png", size: "#{400/2}x#{200/2}", alt: 'AWS EC2 Auto Scaling Group Menu' %>`

**Comment by plaindocs** [comment]
:tada:

File: pages/agent/v3/elastic_ci_aws.md.erb [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+<!-- alex ignore master -->
+
+<a href="https://console.aws.amazon.com/cloudformation/home#/stacks/new?stackName=buildkite&templateURL=https://s3.amazonaws.com/buildkite-aws-stack/latest/aws-stack.yml"><%= image "launch-stack.svg", alt: "Launch stack button" %></a>
+
+<!-- vale on -->    
+
+> The stack creates its own VPC (virtual private cloud) by default. Best practice involves setting up a separate development AWS account and using role switching and consolidated billing. Refer to the [Delegate Access Across AWS Accounts tutorial](http://docs.aws.amazon.com/IAM/latest/UserGuide/tutorial_cross-account-with-roles.html) for more information.
+
+
+If you want to use the [AWS CLI](https://aws.amazon.com/cli/), download [`config.json.example`](https://github.com/buildkite/elastic-ci-stack-for-aws/blob/master/config.json.example), rename it to `config.json`, add your values, and then run the below command:
```
Comment (plaindocs) [inline pages/agent/v3/elastic_ci_aws.md.erb:23]: I'm still slight caught up on this. Should it be "use AWS CLI instead, " ...

And "add your values" is vague. You need to add the Agent Token, are the others required?

File: pages/agent/v3/elastic_ci_aws.md.erb [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+
+If you want to set secrets that your build can access, create a file that sets environment variables and upload it:
+
+```bash
+echo "export MY_ENV_VAR=something secret" > myenv
+aws s3 cp --acl private --sse aws:kms myenv "s3://${SecretsBucket}/env"
+rm myenv
+```
+<div class="Docs__note">
+Currently, you must use the default KMS key for S3. Follow <a href="https://github.com/buildkite/elastic-ci-stack-for-aws/issues/235" target="_blank">issue #235</a> for progress on using specific KMS keys.
```
Comment (plaindocs) [inline pages/agent/v3/elastic_ci_aws.md.erb:89]: As of June 2021 ?

File: pages/agent/v3/elastic_ci_aws.md.erb [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+
+
+* [Amazon Linux 2 LTS](https://aws.amazon.com/amazon-linux-2/)
+* [Buildkite Agent v3.27.0](https://buildkite.com/docs/agent)
+* [Docker](https://www.docker.com) - 19.03.13 (Linux) and 19.03.12 (Windows)
+* [Docker Compose](https://docs.docker.com/compose/) - 1.27.4 (Linux) and 1.27.2 (Windows)
+* [AWS CLI](https://aws.amazon.com/cli/) - useful for performing any ops-related tasks
+<!-- vale off -->
+* [jq](https://stedolan.github.io/jq/) - useful for manipulating JSON responses from CLI tools such as AWS CLI or the Buildkite API
+<!-- vale on -->
```
Comment (plaindocs) [inline pages/agent/v3/elastic_ci_aws.md.erb:104]:
```suggestion
<!-- vale off -->

* [Amazon Linux 2 LTS](https://aws.amazon.com/amazon-linux-2/)
* [Buildkite Agent v3.27.0](https://buildkite.com/docs/agent)
* [Docker](https://www.docker.com) - 19.03.13 (Linux) and 19.03.12 (Windows)
* [Docker Compose](https://docs.docker.com/compose/) - 1.27.4 (Linux) and 1.27.2 (Windows)
* [AWS CLI](https://aws.amazon.com/cli/) - useful for performing any ops-related tasks
* [jq](https://stedolan.github.io/jq/) - useful for manipulating JSON responses from CLI tools such as AWS CLI or the Buildkite API

<!-- vale on -->
```

image

File: pages/agent/v3/elastic_ci_aws.md.erb [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+
+Examples:
+
+* A `docker-builders` stack that provides always-on workers with hot Docker caches (see [Optimizing for slow Docker builds](#optimizing-for-slow-docker-builds))
+* A `pipeline-uploaders` stack with tiny, always-on instances for lightning fast `buildkite-agent pipeline upload` jobs.
+* A `deploy` stack with added credentials and permissions specifically for deployment.
+
+## Autoscaling
+
+If you configure `MinSize` < `MaxSize`, the stack automatically scales up and down based on the number of scheduled jobs.
```
Comment (plaindocs) [inline pages/agent/v3/elastic_ci_aws.md.erb:128]: Is this "configure" in the example config above? or in AWS config? Or either?

File: pages/agent/v3/elastic_ci_aws.md.erb [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+
+This means you can scale down to zero when idle, which means you can use larger instances for the same cost.
+
+Metrics are collected with a Lambda function, polling every minute based on the queue the stack is configured with. The autoscaler monitors only one queue.
```
Comment (plaindocs) [inline pages/agent/v3/elastic_ci_aws.md.erb:132]: > The autoscaler monitors only one queue. 

what are the implications? Can you create a second queue?

File: pages/agent/v3/elastic_ci_aws.md.erb (hunk 10 above)
Comment (plaindocs) [reply pages/agent/v3/elastic_ci_aws.md.erb:128]: Or in buildkite-agent.cfg

File: pages/agent/v3/elastic_ci_aws.md.erb [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+
+We publish the latest build of the stack to `https://s3.amazonaws.com/buildkite-aws-stack/master/aws-stack.yml`, along with a version for each commit in the form of `https://s3.amazonaws.com/buildkite-aws-stack/master/${COMMIT}.aws-stack.yml`.
+
+Branches are published in the form of `https://s3.amazonaws.com/buildkite-aws-stack/${BRANCH}/aws-stack.yml`.
+
+## Updating your stack
+
+To update your stack to the latest version use CloudFormation’s stack update tools with one of the URLs in the [Elastic Stack CI releases](#elastic-stack-ci-releases) section.
+
+Prior to updating, it's a good idea to set the desired instance size on the `AutoscalingGroup` to `0`. manually.
```
Comment (plaindocs) [inline pages/agent/v3/elastic_ci_aws.md.erb:170]:
```suggestion
Prior to updating, it's a good idea to set the desired instance size on the `AutoscalingGroup` to `0` manually.
```

File: pages/agent/v3/elastic_ci_aws.md.erb [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+```json
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Action": [
+        "s3:GetObject"
+      ],
+      "Resource": ["arn:aws:s3:::my_bucket_name/my_bootstrap.sh"]
```
Comment (plaindocs) [inline pages/agent/v3/elastic_ci_aws.md.erb:208]:
```suggestion
      "Resource": ["arn\:aws\:s3::my_bucket_name/my_bootstrap.sh"]

```

File: pages/agent/v3/elastic_ci_aws.md.erb [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+
+1. Create a Docker Hub repository for pushing images to
+1. Update the pipeline’s `env` hook in your secrets bucket to perform a `docker login`
+1. Create a builder stack with its own queue (i.e. `elastic-builders`)
+
+Here is an example build pipeline based on a production Rails application:
+
+```yaml
+steps:
+  - name: ":docker: :package:"
```
Comment (plaindocs) [inline pages/agent/v3/elastic_ci_aws.md.erb:230]:
```suggestion
  - name: "\:docker\: :package:"
```

File: pages/agent/v3/elastic_ci_aws.md.erb [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+    command: ".buildkite/steps/tests"
+    plugins:
+      docker-compose:
+        run: app
+    agents:
+      queue: elastic
+    parallelism: 75
+```
+
+See [Issue 81](https://github.com/buildkite/elastic-ci-stack-for-aws/issues/81) for ideas on other solutions (contributions welcome!).
```
Comment (plaindocs) [inline pages/agent/v3/elastic_ci_aws.md.erb:248]:
```suggestion
See [Issue 81](https://github.com/buildkite/elastic-ci-stack-for-aws/issues/81) for ideas on other solutions (contributions are welcome).
```

**Review by plaindocs — COMMENTED** [review]
This is great, and so close, but I have a few remaining nits and suggestions.

File: pages/agent/v3/elastic_ci_aws.md.erb (hunk 11 above)
Comment (StarfallProjects) [reply pages/agent/v3/elastic_ci_aws.md.erb:134]: @yob could you give me more info here?

**Comment by plaindocs** [comment]
@StarfallProjects maybe we should bump security way to to just after the quickstart? What do you think?

File: pages/agent/v3/elastic_ci_aws.md.erb (hunk 11 above)
Comment (yob) [reply pages/agent/v3/elastic_ci_aws.md.erb:134]: > polling every minute based on the queue the stack is configured with

I think it actually polls every 10s these days, but maybe we can just remove the specificity from the docs?

> The autoscaler monitors only one queue.
>
>> what are the implications? Can you create a second queue?

You can have as many queues as you like on buildkite, but each stack can only monitor a single queue. The montoring is what drives the autoscaling up of the stack, so we only scale up if there's scheduled jobs on one queue.

Maybe that will change one day, but it's not on our short term plans. The implication is that customers will probably want/need one elastic stack per queue.

Comment (StarfallProjects) [reply pages/agent/v3/elastic_ci_aws.md.erb]: Thanks!

**Review by plaindocs — APPROVED** [review]
Great work. :shipit:

**Pull request merged**